cpp/src/stream_compaction/distinct.cu

```diff
@@ -57,23 +57,27 @@ std::unique_ptr<table> distinct(table_view const& input,
   }
 
   auto keys_view = input.select(keys);
-  auto table_ptr = cudf::table_device_view::create(keys_view, stream);
-  auto has_null  = nullate::DYNAMIC{cudf::has_nulls(keys_view)};
-  auto const num_rows{table_ptr->num_rows()};
+  auto preprocessed_keys =
+    cudf::experimental::row::hash::preprocessed_table::create(keys_view, stream);
+  auto has_null = nullate::DYNAMIC{cudf::has_nulls(keys_view)};
+  auto const num_rows{keys_view.num_rows()};
 
   hash_map_type key_map{compute_hash_table_size(num_rows),
                         COMPACTION_EMPTY_KEY_SENTINEL,
                         COMPACTION_EMPTY_VALUE_SENTINEL,
                         detail::hash_table_allocator_type{default_allocator<char>{}, stream},
                         stream.value()};
 
-  compaction_hash hash_key{has_null, *table_ptr};
-  row_equality_comparator row_equal(has_null, *table_ptr, *table_ptr, nulls_equal);
+  auto row_hash = cudf::experimental::row::hash::row_hasher(preprocessed_keys);
+  experimental::compaction_hash hash_key(row_hash.device_hasher(has_null));
+
+  cudf::experimental::row::equality::self_comparator row_equal(preprocessed_keys);
```

Comment:
The new `self_comparator` has been applied in `distinct` but not `distinct_count`. This is problematic since the output of those two APIs should be coherent. Once it's done, `cudf::detail::compaction_hash` in `stream_compaction_common.cuh` can be removed as well. Same for `unique` and `unique_count`.

Reply:
@PointKernel the goal here was to just port a single algorithm to demonstrate the new comparator. Eventually everything will be ported, but not in this PR.

Reply:
On the other hand, I intentionally left this out so that if distinct and distinct count don't match, we know to look for bugs in either comparator

Reply:
Got it. Thank you both for the clarification.